src/lib/parsers/sol-json.c

```diff
@@ -789,6 +790,440 @@ sol_json_serialize_boolean(struct sol_buffer *buffer, bool val)
+        return 0;
+    }
+    case SOL_MEMDESC_TYPE_ENUMERATION: {
+        enum sol_json_type tt;
+
+        tt = sol_json_token_get_type(token);
+        if (tt == SOL_JSON_TYPE_NUMBER) {
+            int64_t v;
+
+            r = sol_json_token_get_int64(token, &v);
+            if (r < 0)
+                return r;
+
+            return sol_memdesc_set_as_int64(desc, memory, v);
+        } else if (tt == SOL_JSON_TYPE_STRING) {
+            struct sol_buffer buf;
+
+            r = sol_json_token_get_unescaped_string(token, &buf);
+            if (r < 0)
+                return r;
+
+            r = sol_memdesc_enumeration_from_str(desc, memory, sol_buffer_get_slice(&buf));
+            sol_buffer_fini(&buf);
+            return r;
+        } else {
+            SOL_WRN("enumerations should be number of string, got json-type %c: %.*s",
+                tt, sol_json_token_get_size(token), token->start);
+            return -EINVAL;
+        }
+    }
+    case SOL_MEMDESC_TYPE_PTR: {
+        enum sol_json_type tt;
+        void **m = memory;
+
+        tt = sol_json_token_get_type(token);
+        if (tt == SOL_JSON_TYPE_NULL) {
+            if (*m && desc->pointed_item)
+                sol_memdesc_free(desc->pointed_item, *m);
+            *m = NULL;
+        } else if (!desc->pointed_item) {
+            uint64_t v;
+
+            r = sol_json_token_get_uint64(token, &v);
+            if (v > ULONG_MAX)
+                return -ERANGE;
+            *m = (void *)(uintptr_t)v;
+        } else {
+            if (!*m) {
+                *m = sol_memdesc_new_with_defaults(desc->pointed_item);
+                if (!*m)
+                    return -ENOMEM;
+            }
+            return sol_json_load_memdesc(token, desc->pointed_item, *m);
+        }
+        return 0;
+    }
+    case SOL_MEMDESC_TYPE_STRUCTURE: {
+        const struct sol_memdesc_structure_member *itr;
+        struct sol_json_scanner scanner;
+        struct sol_json_token sub, key, value;
+        enum sol_json_loop_reason reason;
+        enum sol_json_type tt;
+        bool *done;
+        size_t len, idx;
+        int ret = 0;
+
+        if (!desc->structure_members) {
+            SOL_WRN("desc=%p is SOL_MEMDESC_TYPE_STRUCTURE but does not provide structure_members", desc);
+            return -EINVAL;
+        }
+#ifndef SOL_NO_API_VERSION
+        if (desc->structure_members->base.api_version != SOL_MEMDESC_API_VERSION) {
+            SOL_WRN("structure_members %p->api_version(%" PRIu16 ") != SOL_MEMDESC_API_VERSION (%" PRIu16 ")",
+                desc->structure_members, desc->structure_members->base.api_version, SOL_MEMDESC_API_VERSION);
+            return -EINVAL;
+        }
+#endif
+
+        tt = sol_json_token_get_type(token);
+        if (tt != SOL_JSON_TYPE_OBJECT_START)
+            return -EINVAL;
+
+        len = 0;
+        SOL_MEMDESC_FOREACH_STRUCTURE_MEMBER(desc, itr) {
+            len++;
+        }
+
+        done = alloca(len * sizeof(bool));
```

Comment:
Not bothering with that since the expected use is not to get this from end users, like network, text files or stdin.